### ISSUE TITLE
Allow providing a specification to `pivot()` and `melt()`

### ISSUE THREAD

**Issue by mgirlich** [pola-rs/polars#13834]
### Description

In the R package tidyr there are lower level versions of `pivot_longer/wider()` that allow the user to provide a specification: [`pivot_longer_spec()` and `build_longer_spec()`](https://tidyr.tidyverse.org/reference/pivot_longer_spec.html) and [`pivot_wider_spec()` and `build_wider_spec()`](https://tidyr.tidyverse.org/reference/pivot_wider_spec.html).

They give you much more control on how the pivot should be done.
In particular for `pivot_wider()` this is great as the output columns are known before. This gives type/shape stability of the resulting data frame.
And this would also **allow to pivot in lazy mode**.

**Comment by JulianCologne** [comment]
I am all for it!

Already made a similar issue here:
#11967 -> enable `pivot` on LazyFrame and speed up calculation by providing column values
I really like the argument of the type/shape stability of the resuling DataFrame as without the column specification you have no idea what the resulting df will look like.

also related:
((#11974 ranaming to `pivot_longer`, `pivot_wider`))

**Comment by deanm0000** [comment]
Could you summarize what this means so people can get a sense of it without reading an article or the manual of another library?

**Comment by mgirlich** [comment]
I added the example. Does this make it clear enough or should I add more (though the example is already quite long).

**Comment by deanm0000** [comment]
In my opinion, it'd more accessible to take out all of the R syntax and just stick to polars formatted tables and describing what you want rather than showing the tidyr syntax.

For example:
```
shape: (10, 5)
┌───────┬────────────┬──────────┬──────────┬─────┐
│ GEOID ┆ NAME       ┆ variable ┆ estimate ┆ moe │
│ ---   ┆ ---        ┆ ---      ┆ ---      ┆ --- │
│ i32   ┆ str        ┆ str      ┆ i32      ┆ i32 │
╞═══════╪════════════╪══════════╪══════════╪═════╡
│ 1     ┆ Alabama    ┆ income   ┆ 24476    ┆ 136 │
│ 1     ┆ Alabama    ┆ rent     ┆ 747      ┆ 3   │
│ 2     ┆ Alaska     ┆ income   ┆ 32940    ┆ 508 │
│ 2     ┆ Alaska     ┆ rent     ┆ 1200     ┆ 13  │
│ 4     ┆ Arizona    ┆ income   ┆ 27517    ┆ 148 │
│ 4     ┆ Arizona    ┆ rent     ┆ 972      ┆ 4   │
│ 5     ┆ Arkansas   ┆ income   ┆ 23789    ┆ 165 │
│ 5     ┆ Arkansas   ┆ rent     ┆ 709      ┆ 5   │
│ 6     ┆ California ┆ income   ┆ 29454    ┆ 109 │
│ 6     ┆ California ┆ rent     ┆ 1358     ┆ 3   │
└───────┴────────────┴──────────┴──────────┴─────┘
```

With this people can copy and paste that exact thing and just feed it to `pl.from_repr` if they want to play around with it for a bit.

That aside, I think it'd be good, I am not a big fan of the default names that are output by pivot so this would solve that.

I'm still not a fan of renaming melt and pivot to `pivot_longer` and `pivot_wider` but that's another topic.
